src/bokeh/models/sources.py

```diff
@@ -562,6 +562,11 @@ def _stream(self, new_data: DataDict | pd.Series | pd.DataFrame,
 
         oldkeys = set(self.data.keys())
 
+        # stream on empty source
+        if len(oldkeys) == 0:
+            self.data.update({k: [] for k in newkeys})
+            oldkeys = set(self.data.keys())
```

Comment:
Can we make the next branch an `elif` and skip this update to `oldkeys`? I don't like that other branch being properly skipped is reliant on this otherwise unnecessary assignment happening in just the right place.